mlrun/runtimes/__init__.py

```diff
@@ -12,25 +12,81 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from mlrun import config as mlconfig
 from .base import RunError, BaseRuntime  # noqa
 from .local import HandlerRuntime, LocalRuntime  # noqa
 from .function import RemoteRuntime, new_model_server  # noqa
-from .mpijob import MpiRuntime  # noqa
+from .mpijob import MpiRuntimeV1Alpha1  # noqa
+from .mpijob import MpiRuntimeV1  # noqa
 from .daskjob import DaskCluster, get_dask_resource  # noqa
 from .kubejob import KubejobRuntime  # noqa
 from .sparkjob import SparkRuntime  # noqa
 from .nuclio import nuclio_init_hook
 from .serving import MLModelServer
+from mlrun.k8s_utils import get_k8s_helper
+
+
+class RuntimeKinds(object):
+    remote = 'remote'
+    nuclio = 'nuclio'
+    dask = 'dask'
+    job = 'job'
+    spark = 'spark'
+    mpijob = 'mpijob'
+
+    @staticmethod
+    def get_all():
+        return [RuntimeKinds.remote,
+                RuntimeKinds.nuclio,
+                RuntimeKinds.dask,
+                RuntimeKinds.job,
+                RuntimeKinds.spark,
+                RuntimeKinds.mpijob]
 
 
 runtime_resources_map = {
-    'dask': get_dask_resource()
+    RuntimeKinds.dask: get_dask_resource()
 }
 
 
-runtime_dict = {'remote': RemoteRuntime,
-                'nuclio': RemoteRuntime,
-                'dask': DaskCluster,
-                'job': KubejobRuntime,
-                'mpijob': MpiRuntime,
-                'spark': SparkRuntime}
+def get_runtime_class(kind: str):
+    if kind == RuntimeKinds.mpijob:
+        return _resolve_mpi_runtime()
+
+    kind_runtime_map = {
+        RuntimeKinds.remote: RemoteRuntime,
+        RuntimeKinds.nuclio: RemoteRuntime,
+        RuntimeKinds.dask: DaskCluster,
+        RuntimeKinds.job: KubejobRuntime,
+        RuntimeKinds.spark: SparkRuntime
+    }
+
+    return kind_runtime_map[kind]
+
+
+# resolve mpijob runtime according to the mpi-operator's supported crd-version
+# if specified on mlrun config set it likewise,
+# if not specified, try resolving it according to the mpi-operator, otherwise default to `v1alpha1`
+def _resolve_mpi_runtime():
+
+    # try get mpijob crd version from config
+    mpijob_crd_version = mlconfig.config.try_get('mpijob_crd_version')
+
+    if not mpijob_crd_version:
+        k8s_helper = get_k8s_helper()
+        namespace = k8s_helper.ns()
+
+        # default to v1alpha1 for backwards compatibility
+        mpijob_crd_version = 'v1alpha1'
+
+        # try resolving according to mpi-operator that's running
+        res = k8s_helper.list_pods(namespace=namespace, selector='component=mpi-operator')
+        if len(res) > 0:
+            mpi_operator_pod = res[0]
+            mpijob_crd_version = mpi_operator_pod.metadata.labels.get('crd-version', mpijob_crd_version)
+
+    crd_version_to_runtime = {
+        'v1alpha1': MpiRuntimeV1Alpha1,
+        'v1': MpiRuntimeV1
+    }
+    return crd_version_to_runtime[mpijob_crd_version]
```

Comment:
If an un-supported crd version will be specified by the user it will explode here on a un-understandable _Key Error_
move the crd versions to a consts class, and right after you get it from the config (line 73) verify it's in `MPIJobCRDVersions.all()`